File: apps/api/src/modules/reviews/application/anti-spam.service.ts
import { Injectable } from '@nestjs/common';
import {
  characterDiversity,
  shoutingRatio,
  wordCount,
  type ReviewModerationStatus,
} from '@gamescore/shared';

import { AppConfigService } from '../../../common/config/app-config.service';
import { BadRequestError, TooManyRequestsError } from '../../../common/errors/app.exception';
import { ERROR_CODES } from '../../../common/errors/error-codes';
import { ReviewRepository } from '../repositories/review.repository';

export interface AntiSpamInput {
  userId: string;
  text: string;
  fingerprint: string;
  ipHash: string | null;
}

export interface AntiSpamResult {
  moderationStatus: ReviewModerationStatus;
  rejected: boolean;
}

@Injectable()
export class AntiSpamService {
  constructor(
    private readonly reviews: ReviewRepository,
    private readonly config: AppConfigService,
  ) {}

  async inspect(input: AntiSpamInput): Promise<AntiSpamResult> {
    const { minimumTextLength, maxReviewsPerHour, maxReviewsPerDay, maxAccountsPerIp } =
      this.config.antiSpam;

    if (input.text.trim().length < minimumTextLength) {
      throw new BadRequestError(
        ERROR_CODES.REVIEW_REJECTED_AS_SPAM,
        `Reviews must be at least ${minimumTextLength} characters`,
      );
    }

    if (wordCount(input.text) < 4) {
      throw new BadRequestError(ERROR_CODES.REVIEW_REJECTED_AS_SPAM, 'Review text is too short');
    }

    if (characterDiversity(input.text) < 0.15) {
      throw new BadRequestError(
        ERROR_CODES.REVIEW_REJECTED_AS_SPAM,
        'Review text looks like padding or spam',
      );
    }

    const hourAgo = new Date(Date.now() - 60 * 60 * 1000);
    const dayAgo = new Date(Date.now() - 24 * 60 * 60 * 1000);

    const [hourCount, dayCount, duplicateCount, ipAccounts] = await Promise.all([
      this.reviews.countRecentByUser(input.userId, hourAgo),
      this.reviews.countRecentByUser(input.userId, dayAgo),
      this.reviews.countRecentByFingerprint(input.fingerprint, dayAgo, input.userId),
      input.ipHash ? this.reviews.countDistinctUsersByIp(input.ipHash, dayAgo) : Promise.resolve(0),
    ]);

    if (hourCount >= maxReviewsPerHour || dayCount >= maxReviewsPerDay) {
      throw new TooManyRequestsError(
        ERROR_CODES.REVIEW_RATE_LIMIT_EXCEEDED,
        'You are posting reviews too quickly',
      );
    }

    if (duplicateCount >= 2) {
      throw new BadRequestError(
        ERROR_CODES.REVIEW_REJECTED_AS_SPAM,
        'This review text has already been posted recently',
      );
    }

    let moderationStatus: ReviewModerationStatus = 'NORMAL';

    if (shoutingRatio(input.text) > 0.7 || duplicateCount > 0) {
      moderationStatus = 'SUSPICIOUS';
    }

    if (input.ipHash && ipAccounts >= maxAccountsPerIp) {
      moderationStatus = 'MODERATION_REQUIRED';
    }

    return { moderationStatus, rejected: false };
  }
}

File: apps/api/src/GameScore.Infrastructure/Reviews/AntiSpamService.cs
using GameScore.Application.Configuration;
using GameScore.Application.Errors;
using GameScore.Domain.Moderation;
using GameScore.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Options;

namespace GameScore.Infrastructure.Reviews;

public sealed class AntiSpamService(GameScoreDbContext db, IOptions<AppConfig> config)
{
    public async Task<DbReviewModerationStatus> InspectAsync(
        AntiSpamInput input,
        CancellationToken cancellationToken = default)
    {
        var antiSpam = config.Value.AntiSpam;
        var text = input.Text.Trim();

        if (text.Length < antiSpam.MinimumTextLength)
        {
            throw new BadRequestException(
                ErrorCodes.ReviewRejectedAsSpam,
                $"Reviews must be at least {antiSpam.MinimumTextLength} characters");
        }

        if (TextHelpers.WordCount(text) < 4)
        {
            throw new BadRequestException(ErrorCodes.ReviewRejectedAsSpam, "Review text is too short");
        }

        if (TextHelpers.CharacterDiversity(text) < 0.15)
        {
            throw new BadRequestException(
                ErrorCodes.ReviewRejectedAsSpam,
                "Review text looks like padding or spam");
        }

        var hourAgo = DateTime.UtcNow.AddHours(-1);
        var dayAgo = DateTime.UtcNow.AddDays(-1);

        var hourCount = input.SkipRateLimits
            ? 0
            : await db.Reviews.CountAsync(
                r => r.UserId == input.UserId && r.CreatedAt >= hourAgo && r.DeletedAt == null,
                cancellationToken);

        var dayCount = input.SkipRateLimits
            ? 0
            : await db.Reviews.CountAsync(
                r => r.UserId == input.UserId && r.CreatedAt >= dayAgo && r.DeletedAt == null,
                cancellationToken);

        var duplicateCount = await db.Reviews.CountAsync(
            r =>
                r.TextFingerprint == input.Fingerprint
                && r.CreatedAt >= dayAgo
                && r.UserId != input.UserId
                && (input.ExcludeReviewId == null || r.Id != input.ExcludeReviewId),
            cancellationToken);

        var ipAccounts = !input.SkipRateLimits && input.IpHash is not null
            ? await db.Reviews
                .Where(r => r.AuthorIpHash == input.IpHash && r.CreatedAt >= dayAgo)
                .Select(r => r.UserId)
                .Distinct()
                .CountAsync(cancellationToken)
            : 0;

        if (!input.SkipRateLimits && (hourCount >= antiSpam.MaxReviewsPerHour || dayCount >= antiSpam.MaxReviewsPerDay))
        {
            throw new TooManyRequestsException(
                ErrorCodes.ReviewRateLimitExceeded,
                "You are posting reviews too quickly");
        }

        if (duplicateCount >= 2)
        {
            throw new BadRequestException(
                ErrorCodes.ReviewRejectedAsSpam,
                "This review text has already been posted recently");
        }

        var moderationStatus = DbReviewModerationStatus.NORMAL;
        if (TextHelpers.ShoutingRatio(text) > 0.7 || duplicateCount > 0)
        {
            moderationStatus = DbReviewModerationStatus.SUSPICIOUS;
        }

        if (!input.SkipRateLimits && input.IpHash is not null && ipAccounts >= antiSpam.MaxAccountsPerIp)
        {
            moderationStatus = DbReviewModerationStatus.MODERATION_REQUIRED;
        }

        return moderationStatus;
    }
}

public sealed record AntiSpamInput(
    Guid UserId,
    string Text,
    string Fingerprint,
    string? IpHash,
    bool SkipRateLimits = false,
    Guid? ExcludeReviewId = null);
